// @metaplex-foundation/mpl-bubblegum/dist/src/generated/programs/mplBubblegum.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMplBubblegumProgramId = exports.getMplBubblegumProgram = exports.createMplBubblegumProgram = exports.MPL_BUBBLEGUM_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.MPL_BUBBLEGUM_PROGRAM_ID = 'BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY';
function createMplBubblegumProgram() {
    return {
        name: 'mplBubblegum',
        publicKey: exports.MPL_BUBBLEGUM_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getMplBubblegumErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getMplBubblegumErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createMplBubblegumProgram = createMplBubblegumProgram;
function getMplBubblegumProgram(context, clusterFilter) {
    return context.programs.get('mplBubblegum', clusterFilter);
}
exports.getMplBubblegumProgram = getMplBubblegumProgram;
function getMplBubblegumProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('mplBubblegum', exports.MPL_BUBBLEGUM_PROGRAM_ID, clusterFilter);
}
exports.getMplBubblegumProgramId = getMplBubblegumProgramId;
