// @metaplex-foundation/mpl-account-compression/dist/src/generated/instructions/insertOrAppend.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertOrAppend = exports.getInsertOrAppendInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getInsertOrAppendInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['root', (0, serializers_1.bytes)({ size: 32 })],
        ['leaf', (0, serializers_1.bytes)({ size: 32 })],
        ['index', (0, serializers_1.u32)()],
    ], { description: 'InsertOrAppendInstructionData' }), (value) => ({
        ...value,
        discriminator: [6, 42, 50, 190, 51, 109, 178, 168],
    }));
}
exports.getInsertOrAppendInstructionDataSerializer = getInsertOrAppendInstructionDataSerializer;
// Instruction.
function insertOrAppend(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplAccountCompression', 'mcmt6YrQEMKw8Mw43FmpRLmf7BqRnFMKmAcbxE3xkAW');
    // Accounts.
    const resolvedAccounts = {
        merkleTree: { index: 0, isWritable: true, value: input.merkleTree ?? null },
        authority: { index: 1, isWritable: false, value: input.authority ?? null },
        noop: { index: 2, isWritable: false, value: input.noop ?? null },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getInsertOrAppendInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.insertOrAppend = insertOrAppend;
