// @metaplex-foundation/mpl-bubblegum/dist/src/generated/types/metadataArgs.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMetadataArgsSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const _1 = require(".");
function getMetadataArgsSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['name', (0, serializers_1.string)()],
        ['symbol', (0, serializers_1.string)()],
        ['uri', (0, serializers_1.string)()],
        ['sellerFeeBasisPoints', (0, serializers_1.u16)()],
        ['primarySaleHappened', (0, serializers_1.bool)()],
        ['isMutable', (0, serializers_1.bool)()],
        ['editionNonce', (0, serializers_1.option)((0, serializers_1.u8)())],
        ['tokenStandard', (0, serializers_1.option)((0, _1.getTokenStandardSerializer)())],
        ['collection', (0, serializers_1.option)((0, _1.getCollectionSerializer)())],
        ['uses', (0, serializers_1.option)((0, _1.getUsesSerializer)())],
        ['tokenProgramVersion', (0, _1.getTokenProgramVersionSerializer)()],
        ['creators', (0, serializers_1.array)((0, _1.getCreatorSerializer)())],
    ], { description: 'MetadataArgs' }), (value) => ({
        ...value,
        symbol: value.symbol ?? '',
        primarySaleHappened: value.primarySaleHappened ?? false,
        isMutable: value.isMutable ?? true,
        editionNonce: value.editionNonce ?? (0, umi_1.none)(),
        tokenStandard: value.tokenStandard ?? (0, umi_1.some)(_1.TokenStandard.NonFungible),
        uses: value.uses ?? (0, umi_1.none)(),
        tokenProgramVersion: value.tokenProgramVersion ?? _1.TokenProgramVersion.Original,
    }));
}
exports.getMetadataArgsSerializer = getMetadataArgsSerializer;
