// @metaplex-foundation/mpl-bubblegum/dist/src/generated/types/updateArgs.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUpdateArgsSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const _1 = require(".");
function getUpdateArgsSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['name', (0, serializers_1.option)((0, serializers_1.string)())],
        ['symbol', (0, serializers_1.option)((0, serializers_1.string)())],
        ['uri', (0, serializers_1.option)((0, serializers_1.string)())],
        ['creators', (0, serializers_1.option)((0, serializers_1.array)((0, _1.getCreatorSerializer)()))],
        ['sellerFeeBasisPoints', (0, serializers_1.option)((0, serializers_1.u16)())],
        ['primarySaleHappened', (0, serializers_1.option)((0, serializers_1.bool)())],
        ['isMutable', (0, serializers_1.option)((0, serializers_1.bool)())],
    ], { description: 'UpdateArgs' }), (value) => ({
        ...value,
        name: value.name ?? (0, umi_1.none)(),
        symbol: value.symbol ?? (0, umi_1.none)(),
        uri: value.uri ?? (0, umi_1.none)(),
        creators: value.creators ?? (0, umi_1.none)(),
        sellerFeeBasisPoints: value.sellerFeeBasisPoints ?? (0, umi_1.none)(),
        primarySaleHappened: value.primarySaleHappened ?? (0, umi_1.none)(),
        isMutable: value.isMutable ?? (0, umi_1.none)(),
    }));
}
exports.getUpdateArgsSerializer = getUpdateArgsSerializer;
