// @metaplex-foundation/mpl-bubblegum/dist/src/generated/instructions/collectV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectV2 = exports.getCollectV2InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getCollectV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })]], { description: 'CollectV2InstructionData' }), (value) => ({ ...value, discriminator: [21, 11, 159, 47, 4, 195, 106, 56] }));
}
exports.getCollectV2InstructionDataSerializer = getCollectV2InstructionDataSerializer;
// Instruction.
function collectV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplBubblegum', 'BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    // Accounts.
    const resolvedAccounts = {
        treeConfig: { index: 0, isWritable: true, value: input.treeConfig ?? null },
        destination: {
            index: 1,
            isWritable: true,
            value: input.destination ?? null,
        },
    };
    // Default values.
    if (!resolvedAccounts.destination.value) {
        resolvedAccounts.destination.value = (0, umi_1.publicKey)('2dgJVPC5fjLTBTmMvKDRig9JJUGK2Fgwr3EHShFxckhv');
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getCollectV2InstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.collectV2 = collectV2;
