// @metaplex-foundation/mpl-bubblegum/dist/src/generated/instructions/setAndVerifyCollection.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setAndVerifyCollection = exports.getSetAndVerifyCollectionInstructionDataSerializer = void 0;
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getSetAndVerifyCollectionInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['root', (0, serializers_1.bytes)({ size: 32 })],
        ['dataHash', (0, serializers_1.bytes)({ size: 32 })],
        ['creatorHash', (0, serializers_1.bytes)({ size: 32 })],
        ['nonce', (0, serializers_1.u64)()],
        ['index', (0, serializers_1.u32)()],
        ['metadata', (0, types_1.getMetadataArgsSerializer)()],
        ['collection', (0, serializers_1.publicKey)()],
    ], { description: 'SetAndVerifyCollectionInstructionData' }), (value) => ({
        ...value,
        discriminator: [235, 242, 121, 216, 158, 234, 180, 234],
    }));
}
exports.getSetAndVerifyCollectionInstructionDataSerializer = getSetAndVerifyCollectionInstructionDataSerializer;
// Instruction.
function setAndVerifyCollection(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplBubblegum', 'BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    // Accounts.
    const resolvedAccounts = {
        treeConfig: {
            index: 0,
            isWritable: false,
            value: input.treeConfig ?? null,
        },
        leafOwner: { index: 1, isWritable: false, value: input.leafOwner ?? null },
        leafDelegate: {
            index: 2,
            isWritable: false,
            value: input.leafDelegate ?? null,
        },
        merkleTree: { index: 3, isWritable: true, value: input.merkleTree ?? null },
        payer: { index: 4, isWritable: false, value: input.payer ?? null },
        treeCreatorOrDelegate: {
            index: 5,
            isWritable: false,
            value: input.treeCreatorOrDelegate ?? null,
        },
        collectionAuthority: {
            index: 6,
            isWritable: false,
            value: input.collectionAuthority ?? null,
        },
        collectionAuthorityRecordPda: {
            index: 7,
            isWritable: false,
            value: input.collectionAuthorityRecordPda ?? null,
        },
        collectionMint: {
            index: 8,
            isWritable: false,
            value: input.collectionMint ?? null,
        },
        collectionMetadata: {
            index: 9,
            isWritable: true,
            value: input.collectionMetadata ?? null,
        },
        collectionEdition: {
            index: 10,
            isWritable: false,
            value: input.collectionEdition ?? null,
        },
        bubblegumSigner: {
            index: 11,
            isWritable: false,
            value: input.bubblegumSigner ?? null,
        },
        logWrapper: {
            index: 12,
            isWritable: false,
            value: input.logWrapper ?? null,
        },
        compressionProgram: {
            index: 13,
            isWritable: false,
            value: input.compressionProgram ?? null,
        },
        tokenMetadataProgram: {
            index: 14,
            isWritable: false,
            value: input.tokenMetadataProgram ?? null,
        },
        systemProgram: {
            index: 15,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.treeConfig.value) {
        resolvedAccounts.treeConfig.value = (0, accounts_1.findTreeConfigPda)(context, {
            merkleTree: (0, shared_1.expectPublicKey)(resolvedAccounts.merkleTree.value),
        });
    }
    if (!resolvedAccounts.leafDelegate.value) {
        resolvedAccounts.leafDelegate.value = (0, shared_1.expectSome)(resolvedAccounts.leafOwner.value);
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.treeCreatorOrDelegate.value) {
        resolvedAccounts.treeCreatorOrDelegate.value = context.identity;
    }
    if (!resolvedAccounts.collectionAuthority.value) {
        resolvedAccounts.collectionAuthority.value = context.identity;
    }
    if (!resolvedAccounts.collectionMetadata.value) {
        resolvedAccounts.collectionMetadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
        });
    }
    if (!resolvedAccounts.collectionEdition.value) {
        resolvedAccounts.collectionEdition.value = (0, mpl_token_metadata_1.findMasterEditionPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value),
        });
    }
    if (!resolvedAccounts.bubblegumSigner.value) {
        resolvedAccounts.bubblegumSigner.value = (0, umi_1.publicKey)('BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    }
    if (!resolvedAccounts.logWrapper.value) {
        resolvedAccounts.logWrapper.value = context.programs.getPublicKey('splNoop', 'noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV');
        resolvedAccounts.logWrapper.isWritable = false;
    }
    if (!resolvedAccounts.compressionProgram.value) {
        resolvedAccounts.compressionProgram.value = context.programs.getPublicKey('splAccountCompression', 'cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK');
        resolvedAccounts.compressionProgram.isWritable = false;
    }
    if (!resolvedAccounts.tokenMetadataProgram.value) {
        resolvedAccounts.tokenMetadataProgram.value = (0, umi_1.publicKey)('BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedArgs.dataHash) {
        resolvedArgs.dataHash = (0, hooked_1.resolveDataHash)(context, resolvedAccounts, resolvedArgs, programId, false);
    }
    if (!resolvedArgs.creatorHash) {
        resolvedArgs.creatorHash = (0, hooked_1.resolveCreatorHash)(context, resolvedAccounts, resolvedArgs, programId, false);
    }
    if (!resolvedArgs.collection) {
        resolvedArgs.collection = (0, shared_1.expectPublicKey)(resolvedAccounts.collectionMint.value);
    }
    if (!resolvedArgs.proof) {
        resolvedArgs.proof = [];
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Remaining Accounts.
    const remainingAccounts = resolvedArgs.proof.map((value, index) => ({
        index,
        value,
        isWritable: false,
    }));
    orderedAccounts.push(...remainingAccounts);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSetAndVerifyCollectionInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.setAndVerifyCollection = setAndVerifyCollection;
