// @metaplex-foundation/mpl-bubblegum/dist/src/generated/instructions/setDecompressibleState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setDecompressibleState = exports.getSetDecompressibleStateInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getSetDecompressibleStateInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['decompressableState', (0, types_1.getDecompressibleStateSerializer)()],
    ], { description: 'SetDecompressibleStateInstructionData' }), (value) => ({
        ...value,
        discriminator: [82, 104, 152, 6, 149, 111, 100, 13],
    }));
}
exports.getSetDecompressibleStateInstructionDataSerializer = getSetDecompressibleStateInstructionDataSerializer;
// Instruction.
function setDecompressibleState(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplBubblegum', 'BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    // Accounts.
    const resolvedAccounts = {
        treeConfig: { index: 0, isWritable: true, value: input.treeConfig ?? null },
        treeCreator: {
            index: 1,
            isWritable: false,
            value: input.treeCreator ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.treeCreator.value) {
        resolvedAccounts.treeCreator.value = context.identity;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSetDecompressibleStateInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.setDecompressibleState = setDecompressibleState;
