// @metaplex-foundation/mpl-bubblegum/dist/src/generated/instructions/unverifyCreator.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.unverifyCreator = exports.getUnverifyCreatorInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getUnverifyCreatorInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['root', (0, serializers_1.bytes)({ size: 32 })],
        ['dataHash', (0, serializers_1.bytes)({ size: 32 })],
        ['creatorHash', (0, serializers_1.bytes)({ size: 32 })],
        ['nonce', (0, serializers_1.u64)()],
        ['index', (0, serializers_1.u32)()],
        ['metadata', (0, types_1.getMetadataArgsSerializer)()],
    ], { description: 'UnverifyCreatorInstructionData' }), (value) => ({
        ...value,
        discriminator: [107, 178, 57, 39, 105, 115, 112, 152],
    }));
}
exports.getUnverifyCreatorInstructionDataSerializer = getUnverifyCreatorInstructionDataSerializer;
// Instruction.
function unverifyCreator(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplBubblegum', 'BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    // Accounts.
    const resolvedAccounts = {
        treeConfig: {
            index: 0,
            isWritable: false,
            value: input.treeConfig ?? null,
        },
        leafOwner: { index: 1, isWritable: false, value: input.leafOwner ?? null },
        leafDelegate: {
            index: 2,
            isWritable: false,
            value: input.leafDelegate ?? null,
        },
        merkleTree: { index: 3, isWritable: true, value: input.merkleTree ?? null },
        payer: { index: 4, isWritable: false, value: input.payer ?? null },
        creator: { index: 5, isWritable: false, value: input.creator ?? null },
        logWrapper: {
            index: 6,
            isWritable: false,
            value: input.logWrapper ?? null,
        },
        compressionProgram: {
            index: 7,
            isWritable: false,
            value: input.compressionProgram ?? null,
        },
        systemProgram: {
            index: 8,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.treeConfig.value) {
        resolvedAccounts.treeConfig.value = (0, accounts_1.findTreeConfigPda)(context, {
            merkleTree: (0, shared_1.expectPublicKey)(resolvedAccounts.merkleTree.value),
        });
    }
    if (!resolvedAccounts.leafDelegate.value) {
        resolvedAccounts.leafDelegate.value = (0, shared_1.expectSome)(resolvedAccounts.leafOwner.value);
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.logWrapper.value) {
        resolvedAccounts.logWrapper.value = context.programs.getPublicKey('splNoop', 'noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV');
        resolvedAccounts.logWrapper.isWritable = false;
    }
    if (!resolvedAccounts.compressionProgram.value) {
        resolvedAccounts.compressionProgram.value = context.programs.getPublicKey('splAccountCompression', 'cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK');
        resolvedAccounts.compressionProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedArgs.dataHash) {
        resolvedArgs.dataHash = (0, hooked_1.resolveDataHash)(context, resolvedAccounts, resolvedArgs, programId, false);
    }
    if (!resolvedArgs.creatorHash) {
        resolvedArgs.creatorHash = (0, hooked_1.resolveCreatorHash)(context, resolvedAccounts, resolvedArgs, programId, false);
    }
    if (!resolvedArgs.proof) {
        resolvedArgs.proof = [];
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Remaining Accounts.
    const remainingAccounts = resolvedArgs.proof.map((value, index) => ({
        index,
        value,
        isWritable: false,
    }));
    orderedAccounts.push(...remainingAccounts);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getUnverifyCreatorInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.unverifyCreator = unverifyCreator;
