// @metaplex-foundation/spl-account-compression/dist/src/generated/programs/splAccountCompression.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSplAccountCompressionProgramId = exports.getSplAccountCompressionProgram = exports.createSplAccountCompressionProgram = exports.SPL_ACCOUNT_COMPRESSION_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.SPL_ACCOUNT_COMPRESSION_PROGRAM_ID = 'cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK';
function createSplAccountCompressionProgram() {
    return {
        name: 'splAccountCompression',
        publicKey: exports.SPL_ACCOUNT_COMPRESSION_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getSplAccountCompressionErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getSplAccountCompressionErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createSplAccountCompressionProgram = createSplAccountCompressionProgram;
function getSplAccountCompressionProgram(context, clusterFilter) {
    return context.programs.get('splAccountCompression', clusterFilter);
}
exports.getSplAccountCompressionProgram = getSplAccountCompressionProgram;
function getSplAccountCompressionProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('splAccountCompression', exports.SPL_ACCOUNT_COMPRESSION_PROGRAM_ID, clusterFilter);
}
exports.getSplAccountCompressionProgramId = getSplAccountCompressionProgramId;
