// @metaplex-foundation/spl-account-compression/dist/src/generated/programs/splNoop.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSplNoopProgramId = exports.getSplNoopProgram = exports.createSplNoopProgram = exports.SPL_NOOP_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.SPL_NOOP_PROGRAM_ID = 'noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV';
function createSplNoopProgram() {
    return {
        name: 'splNoop',
        publicKey: exports.SPL_NOOP_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getSplNoopErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getSplNoopErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createSplNoopProgram = createSplNoopProgram;
function getSplNoopProgram(context, clusterFilter) {
    return context.programs.get('splNoop', clusterFilter);
}
exports.getSplNoopProgram = getSplNoopProgram;
function getSplNoopProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('splNoop', exports.SPL_NOOP_PROGRAM_ID, clusterFilter);
}
exports.getSplNoopProgramId = getSplNoopProgramId;
