// @metaplex-foundation/mpl-account-compression/dist/src/generated/programs/mplNoop.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMplNoopProgramId = exports.getMplNoopProgram = exports.createMplNoopProgram = exports.MPL_NOOP_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.MPL_NOOP_PROGRAM_ID = 'mnoopTCrg4p8ry25e4bcWA9XZjbNjMTfgYVGGEdRsf3';
function createMplNoopProgram() {
    return {
        name: 'mplNoop',
        publicKey: exports.MPL_NOOP_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getMplNoopErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getMplNoopErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createMplNoopProgram = createMplNoopProgram;
function getMplNoopProgram(context, clusterFilter) {
    return context.programs.get('mplNoop', clusterFilter);
}
exports.getMplNoopProgram = getMplNoopProgram;
function getMplNoopProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('mplNoop', exports.MPL_NOOP_PROGRAM_ID, clusterFilter);
}
exports.getMplNoopProgramId = getMplNoopProgramId;
